src/txmempool.cpp

```diff
@@ -1307,13 +1345,18 @@ size_t CTxMemPool::_DynamicMemoryUsage() const
            cachedInnerUsage;
 }
 
-void CTxMemPool::_RemoveStaged(setEntries &stage, bool updateDescendants)
+void CTxMemPool::_RemoveStaged(setEntries &stage,
+    bool updateDescendants,
+    CTxMemPool::TxMempoolOriginalStateMap *changeSet)
 {
     {
         AssertWriteLockHeld(cs_txmempool);
-        _UpdateForRemoveFromMempool(stage, updateDescendants);
+        _UpdateForRemoveFromMempool(stage, updateDescendants, changeSet);
         for (const txiter &it : stage)
         {
+            // If a different tx commitment added this one remove it now because its confirmed
+            if (changeSet)
+                changeSet->erase(it);
             removeUnchecked(it);
         }
```

Comment:
aren't you invalidating the iterator here and then trying to access it on the next line...same also with the loop, should maybe use a while loop instead of for()...

changeSet->erase(it)  then on the following line removeUnchecked(it)

Reply:
this is confusing because these are containers of iterators.  The iterator of setEntries is never invalidated so the for loop is correct.
changeSet->erase(it) removes the "it" entry from a completely separate container (changeSet).
Then removeUnchecked removes the "it" entry from the container that it is located in (the mempool), so further uses of "it" would fail.  removeUnchecked(it) needs to be the last line of the loop.